### PR TITLE
Add new LXC: Tianji

### PR DESCRIPTION
**Pull request by MickLesk**
## Description
Add new LXC Tianji.
We use this script in our company to track our build pipelines and develop portal.
The Script runs since 1 Week. 
But you are welcome to give it “free” for testing. My tests are now complete and so far everything is working.

As additional information:
The initial build of the LXC takes about 10-15 minutes, because pnpm sometimes reacts very slowly. The update is significantly faster. 

The initial login is **admin:admin**

image

image

## Type of change
- [x] New script (a fully functional and thoroughly tested script or set of scripts.)
- [x] Self-review performed (I have reviewed my code, ensuring it follows established patterns and conventions)
- [x] Documentation update required (this change requires an update to the documentation)

### CODE REVIEW

**Comment by tteck** [comment]
How does it differ from Uptime Kuma?

**Comment by MickLesk** [comment]
Tianji is better, faster and have more functions than uptime Kuma. Its a mix of uptime Kuma, Nagios, a little bit Prometheus and so on.


Further highlights or descriptions: 
 * More comprehensive: Offers website analytics and server status monitoring in addition to uptime monitoring.

 * All-in-one: Can replace multiple tools, such as Google Analytics or Prometheus.

 * Scalable: Uses a scalable database and is suitable for larger applications.

File: ct/tianji-install.sh [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+source /dev/stdin <<< "$FUNCTIONS_FILE_PATH"
+color
+verb_ip6
+catch_errors
+setting_up_container
+network_check
+update_os
+
+msg_info "Installing Dependencies (Patience)"
+$STD apt-get install -y --no-install-recommends \
```
Comment (tteck) [inline ct/tianji-install.sh]: why do you continue to use `--no-install-recommends`?

File: ct/tianji-install.sh [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  postgresql \
+  build-essential \
+  curl \
+  unzip \
```
Comment (tteck) [inline ct/tianji-install.sh]: not needed if `--no-install-recommends` is removed. `mc` will install unzip

File: ct/tianji-install.sh [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+msg_info "Installing Node.js, pnpm & pm2"
+$STD apt-get install -y nodejs
+$STD npm install -g pnpm@9.7.1
+$STD npm install -g pm2 
+msg_ok "Installed Node.js, pnpm & pm2"
+
+
+msg_info "Setup Tianji (Patience)"
+cd /opt
+RELEASE=$(wget -q https://github.com/msgbyte/tianji/releases/latest -O - | grep "title>Release" | cut -d " " -f 4)
```
Comment (tteck) [inline ct/tianji-install.sh]: `RELEASE=$(curl -s https://api.github.com/repos/msgbyte/tianji/releases/latest | grep "tag_name" | awk '{print substr($2, 3, length($2)-4) }')`

File: ct/tianji-install.sh [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+wget -q "https://github.com/msgbyte/tianji/archive/refs/tags/${RELEASE}.zip"
```
Comment (tteck) [inline ct/tianji-install.sh]: `wget -q "https://github.com/msgbyte/tianji/archive/refs/tags/v${RELEASE}.zip"`

File: ct/tianji-install.sh [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+unzip -q ${RELEASE}.zip
```
Comment (tteck) [inline ct/tianji-install.sh]: `unzip -q v${RELEASE}.zip`

File: ct/tianji-install.sh [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+CLEAN_RELEASE=$(echo $RELEASE | sed 's/^v//')
```
Comment (tteck) [inline ct/tianji-install.sh]: remove

File: ct/tianji-install.sh [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+mv tianji-${CLEAN_RELEASE} /opt/tianji
```
Comment (tteck) [inline ct/tianji-install.sh]: `mv tianji-${RELEASE} /opt/tianji`

File: ct/tianji-install.sh [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+cd tianji
+export NODE_OPTIONS=--max_old_space_size=4096
+echo "${RELEASE}" >"/opt/${APPLICATION}_version.txt"
```
Comment (tteck) [inline ct/tianji-install.sh]: move below `$STD pnpm build`

File: ct/tianji-install.sh [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+$STD pm2 save
+msg_ok "Service activated"
+
+motd_ssh
+customize
+
+msg_info "Cleaning up"
+rm -R /opt/${RELEASE}.zip
+$STD apt-get autoremove
+$STD apt-get autoclean
```
Comment (tteck) [inline ct/tianji-install.sh]: add flag `-y` to both

File: ct/tianji.sh [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  echo_default
+}
+function update_script() {
+header_info
+if [[ ! -d /opt/tianji ]]; then msg_error "No ${APP} Installation Found!"; exit; fi
+if (( $(df /boot | awk 'NR==2{gsub("%","",$5); print $5}') > 80 )); then
+  read -r -p "Warning: Storage is dangerously low, continue anyway? <y/N> " prompt
+  [[ ${prompt,,} =~ ^(y|yes)$ ]] || exit
+fi
+RELEASE=$(wget -q https://github.com/msgbyte/tianji/releases/latest -O - | grep "title>Release" | cut -d " " -f 4)
```
Comment (tteck) [inline ct/tianji.sh]: `RELEASE=$(curl -s https://api.github.com/repos/msgbyte/tianji/releases/latest | grep "tag_name" | awk '{print substr($2, 3, length($2)-4) }')`

File: ct/tianji.sh [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_version.txt)" ]]; then
+
+  msg_info "Stopping ${APP} Service"
+  pm2 stop tianji >/dev/null 2>&1
+  msg_ok "Stopped ${APP} Service"
+
+  msg_info "Updating ${APP} to ${RELEASE}"
+  cd /opt/tianji
+  git checkout -q $RELEASE
```
Comment (tteck) [inline ct/tianji.sh]: `git checkout -q v${RELEASE}`

**Comment by MickLesk** [comment]
Okay i fix some little things, i merge it to my fork and test it. 
I give feedback asap.

File: ct/tianji-install.sh (hunk 8 above)
Comment (tteck) [reply ct/tianji-install.sh]: `echo "${RELEASE}" >"/opt/${APPLICATION}_version.txt"` was not moved

Comment (tteck) [reply ct/tianji-install.sh]: Sorry, yes it was

**Comment by MickLesk** [comment]
so now it works fine & testet with actual script state

**Pull request merged**